network/src/protocols/network/mod.rs

```diff
@@ -172,22 +174,58 @@ pub trait NewNetworkEvents {
     fn new(
         peer_mgr_notifs_rx: aptos_channel::Receiver<(PeerId, ProtocolId), PeerManagerNotification>,
         connection_notifs_rx: aptos_channel::Receiver<PeerId, ConnectionNotification>,
+        bounded_executor: BoundedExecutor,
     ) -> Self;
 }
 
-impl<TMessage: Message> NewNetworkEvents for NetworkEvents<TMessage> {
+impl<TMessage: Message + Send + 'static> NewNetworkEvents for NetworkEvents<TMessage> {
     fn new(
-        peer_mgr_notifs_rx: aptos_channel::Receiver<(PeerId, ProtocolId), PeerManagerNotification>,
+        mut peer_mgr_notifs_rx: aptos_channel::Receiver<
+            (PeerId, ProtocolId),
+            PeerManagerNotification,
+        >,
         connection_notifs_rx: aptos_channel::Receiver<PeerId, ConnectionNotification>,
+        bounded_executor: BoundedExecutor,
     ) -> Self {
-        let data_event_stream = peer_mgr_notifs_rx.filter_map(
-            peer_mgr_notif_to_event
-                as fn(PeerManagerNotification) -> future::Ready<Option<Event<TMessage>>>,
-        );
+        // Create a channel for deserialized messages
+        let (deserialized_message_sender, deserialized_message_receiver) =
+            mpsc::channel(MAX_DESERIALIZATION_QUEUE_SIZE);
+
+        // Deserialize the peer manager notifications in parallel (for each
+        // network application) and send them to the receiver. Note: this
+        // may cause out of order message delivery.
```

Comment:
Would "out of order delivery" create any problem?

Reply:
I'm pretty sure the order matters unless the downstream application handles out of order messages.

The RPC protocol is built around ensuring that the messages are properly in order and most of the applications are expecting that they get the correct message during the RPC call

Reply:
> Would "out of order delivery" create any problem?

I double checked with @zekun000 and @bchocho and we all agreed that it shouldn't. Applications should not be assuming ordering of messages because that assumption can be trivially broken at any point in the stack. I also had a quick look at various components and saw this not to be the case. All applications seem to handle this already (as they should) 😄